src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,188 @@
+            add(value);
+        } else {
+            linkBefore(value, getNode(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        addAll(size, list);
+    }
+
+    private void addAll(int index, Collection<? extends T> collection) {
+        checkAddingIndex(index);
+        T[] objects = (T[]) collection.toArray();
+
+        if (index == size) {
+            for (T elem : objects) {
+                add(elem);
+            }
+        } else {
+            for (T element : objects) {
+                linkBefore(element, getNode(index));
+                index++;
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> setNode = getNode(index);
+        final T oldValue = setNode.item;
+        setNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> current = head; current != null; current = current.next) {
+            if ((current.item != null && current.item.equals(object)) || current.item == object) {
+                unlink(current);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkBefore(T value, Node<T> current) {
+        if (current == null) {
+            throw new IndexOutOfBoundsException("Can't link element "
+                    + value + " before null Node " + current);
+        }
+
+        final Node<T> precedent = current.prev;
+        final Node<T> newNode = new Node<>(precedent, value, current);
+        current.prev = newNode;
+
+        if (precedent == null) {
+            head = newNode;
+        } else {
+            precedent.next = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        checkElementIndex(index);
+
+        if (index < (size / 2)) {
+            Node<T> result = head;
+            for (int i = 0; i < index; i++) {
+                result = result.next;
+            }
+            return result;
+        } else {
+            Node<T> result = tail;
+            for (int i = size - 1; i > index; i--) {
+                result = result.prev;
+            }
+            return result;
+        }
+    }
+
+    private T unlink(Node<T> removeNode) {
+        if (removeNode == null) {
+            throw new NullPointerException("Can't remove the non-exist element " + removeNode);
+        }
```

Comment:
```suggestion
```

Reply:
> Suggested change
>         if (removeNode == null) {
>             throw new NullPointerException("Can't remove the non-exist element " + removeNode);
>         }

I really do not understand why we need to remove null-checking.